neutral_generation/smart_convert.py

```diff
@@ -0,0 +1,284 @@
+import re
+import math
+import torch
+import spacy
+
+# SpaCy: lowercase is for dependency parser, uppercase is for part-of-speech tagger
+from spacy.symbols import nsubj, aux, conj, neg, VERB, AUX
+from pytorch_pretrained_bert import OpenAIGPTTokenizer, OpenAIGPTModel, OpenAIGPTLMHeadModel
+from constants import *
+
+nlp = spacy.load("en_core_web_sm")
+
+# Load pre-trained model (weights)
+model = OpenAIGPTLMHeadModel.from_pretrained('openai-gpt')
+model.eval()
+# Load pre-trained model tokenizer (vocabulary)
+tokenizer = OpenAIGPTTokenizer.from_pretrained('openai-gpt')
+
+
+def convert(s):
+    # input: SNAPE sentence (meaning 1 entity and 1 gender)
+    # output: sentence in gender-neutral form
+
+    # pluralize verbs: change all verbs in third-person singular to third-person plural
+    auxiliaries = identify_verbs_and_auxiliaries(s)
+    s = pluralize_verbs(s, auxiliaries)
+
+    # use regex to replace pronouns and gendered terms that have a clear mapping
+    SIMPLE_REPLACE = EASY_PRONOUNS
+    SIMPLE_REPLACE.update(GENDERED_TERMS)
+    for p, r in SIMPLE_REPLACE.items():
+        s = regex_token_replace(s, p, r)
```

Comment:
This segment here can be profiled and optimized. Have you tried to see if the regex takes a lot of time?

If so, we can flip it. i.e loop through each token in the input and then in the inner loop go over all possible replacements. That might be faster.

Reply:
Yes, that's a good suggestion. Done.